src/GafferScene/Instancer.cpp

```diff
@@ -269,6 +492,129 @@ class Instancer::EngineData : public Data
 			return result;
 		}
 
+		typedef std::vector< boost::unordered_set< IECore::MurmurHash > > PrototypeHashes;
+
+		// In order to compute the number of variations, we compute a unique hash for every context we use
+		// for evaluating prototypes.  So that we can track which sources are responsible for variations,
+		// we return a vector of hash sets, corresponding to the order of m_prototypeContextVariables,
+		// plus an extra entry at the end for the combined result of all variation sources
+		std::shared_ptr<PrototypeHashes> uniquePrototypeHashes() const
+		{
+			std::shared_ptr<PrototypeHashes> result( new PrototypeHashes( m_prototypeContextVariables.size() + 1 ) );
+
+			std::vector< IECore::MurmurHash> prototypeHashes;
+			prototypeHashes.reserve( m_prototypeIndexRemap.size() );
+			for( int indexRemap : m_prototypeIndexRemap )
+			{
+				IECore::MurmurHash h;
+				// We throw away rootStrings during initPrototypes, necessitating interating
+				// all the path components here - if perf was important, it would be better to
+				// store rootStrings
+				InternedStringVectorDataPtr path = m_roots[ indexRemap ];
+				h.append( &(path->readable())[0], path->readable().size() );
+				prototypeHashes.push_back( h );
+			}
+
+
+			size_t n = numPoints();
+			std::vector< IECore::MurmurHash > contextVariableHashes( m_prototypeContextVariables.size() );
+			for( unsigned int i = 0; i < n; i++ )
+			{
+				IECore::MurmurHash totalHash;
+				if( m_indices )
+				{
+					totalHash = prototypeHashes[ ((*m_indices)[i]) % m_numPrototypes ];
+				}
+				hashPrototypeContext( i, contextVariableHashes );
+				for( unsigned int j = 0; j < contextVariableHashes.size(); j++ )
+				{
+					(*result)[j].insert( contextVariableHashes[j] );
+					totalHash.append( contextVariableHashes[j] );
+				}
```

Comment:
The coupling to `hashPrototypeContext()` seems slightly odd. We make a vector, `hashPrototypeContext()` iterates to manipulate that vector, and then we iterate to get the results out of the vector. Any reason not to replace `hashPrototypeContext()` with `hashPrototypeContextVariable()`, to operate on one variable, and then have a single loop on the outside? Maybe  because you're trying to keep `hashPrototypeContext()` aligned in structure to `fillPrototypeContext()`?

Reply:
Yeah, you're right that I was trying to match fillPrototypeContext(), but it doesn't matter much, I've switched it.